eslint.config.mjs

```diff
@@ -0,0 +1,186 @@
+    rules: {
+      'import/export': 'off', // Mostly handled by Typescript itself. ESLint produces false positives with declaration merging.
+      'no-restricted-imports': [
+        'error',
+        {
+          patterns: NO_RESTRICTED_IMPORTS_PATTERNS_DEEPLY_NESTED,
+        },
+      ],
+      '@typescript-eslint/no-redeclare': 'off',
+      // We LOVE non-breaking spaces, and both straight and curly quotes here
+      'no-irregular-whitespace': [1, { skipJSXText: true, skipStrings: true }],
+      'react/no-unescaped-entities': [1, { forbid: ['>', '}'] }],
+      'react/prop-types': 'off',
+      'react-hooks/exhaustive-deps': [
+        'error',
+        {
+          additionalHooks: 'useModernLayoutEffect',
+        },
+      ],
+      'react/no-unused-prop-types': 'off',
+      'material-ui/straight-quotes': 'off',
+      // This prevents us from creating components like `<h1 {...props} />`
+      'jsx-a11y/heading-has-content': 'off',
+      'jsx-a11y/anchor-has-content': 'off',
+
+      // This rule doesn't recognise <label> wrapped around custom controls
+      'jsx-a11y/label-has-associated-control': 'off',
+
+      // An overzealous rule that shouts at <a href="#"> in demos.
+      'jsx-a11y/anchor-is-valid': 'off',
+    },
+  },
+  {
+    files: [
+      // matching the pattern of the test runner
+      '**/*.test.?(c|m)[jt]s?(x)',
+    ],
+    extends: createTestConfig(),
+    rules: {
+      'mocha/no-top-level-hooks': 'off', // not needed
+      // disable eslint-plugin-jsx-a11y
+      // tests are not driven by assistive technology
+      // add `jsx-a11y` rules once you encounter them in tests
+      'jsx-a11y/click-events-have-key-events': 'off',
+      'jsx-a11y/control-has-associated-label': 'off',
+      'jsx-a11y/iframe-has-title': 'off',
+      'jsx-a11y/label-has-associated-control': 'off',
+      'jsx-a11y/mouse-events-have-key-events': 'off',
+      'jsx-a11y/no-noninteractive-tabindex': 'off',
+      'jsx-a11y/no-static-element-interactions': 'off',
+      'jsx-a11y/tabindex-no-positive': 'off',
+
+      // In tests this is generally intended.
+      'react/button-has-type': 'off',
+
+      // disable some mocha rules
+      'mocha/no-async-suite': 'off', // not needed
+      'mocha/no-identical-title': 'off', // not needed
+    },
+  },
+  baseSpecRules,
+  {
+    files: ['packages/**/*.test{.tsx,.js}'],
+    rules: {
+      'testing-library/prefer-screen-queries': 'off', // TODO: enable and fix
+      'testing-library/no-container': 'off', // TODO: enable and fix
+      'testing-library/render-result-naming-convention': 'off', // False positives
+    },
+  },
+  {
+    name: 'MUI ESLint config for docs',
+    files: ['docs/**/*'],
+    extends: createDocsConfig(),
+    rules: {
+      'no-restricted-imports': [
+        'error',
+        {
+          paths: NO_RESTRICTED_IMPORTS_PATHS_TOP_LEVEL_PACKAGES,
+          patterns: NO_RESTRICTED_IMPORTS_PATTERNS_DEEPLY_NESTED,
+        },
+      ],
+    },
+  },
+  {
```

Comment:
One way we could clean this up is to put local config closer to where it should apply.

```tsx
// eslint.config.js
import experiments from 'docs/src/app/(private)/experiments/eslint.config.js';

export default [
  ...experiments
]
```

```tsx
// docs/src/app/(private)/experiments/eslint.config.js
import { getWorkspaceRelative } from '@mui/internal-code-infra';
const ROOT = getWorkspaceRelative(import.meta.dirname);
export default [
  {
    files: [`${ROOT}/**/*{.tsx,.js}`],
    rules: {
      '@typescript-eslint/no-use-before-define': 'off',
      'no-alert': 'off',
      'no-console': 'off',
      'import/no-relative-packages': 'off',
    },
  }
]
```

Perhaps for a follow-up if there is enough interest in this pattern?